rfc/build-system/build-system.md

```diff
@@ -0,0 +1,711 @@
+  tests
+
+This document does not describe any changes to the end-to-end tests, aside from
+how they interact with the build system. While we would like to eventually move
+these tests into the GitHub Actions CI system and retire Jenkins, that will be a
+separate RFC.
+
+It also does not describe any major changes to the way Deck is built. We will
+continue to build it with Gradle using the same infrastructure as all the other
+microservices. Since Deck is markedly different from the rest of the
+microservices, changing its build system
+[has been a topic of discussion recently](https://docs.google.com/document/d/1ch0wEUNlQwvjAn8ob_HPZ54jYbmmloe2AKbtMRI_JsU/edit?usp=sharing),
+but is not in scope for this document.
+
+It doesn't discuss changes to the Halyard and Spin release process.
+
+## Motivation and Rationale
+
+Currently there are two separate build systems for Spinnaker. These systems
+build the same code, but publish their artifacts to different artifact
+repositories using different version numbers. This is redundant and confusing.
+This document aims to describe a single build system for all community
+artifacts.
+
+One system is orchestrated via GitHub Actions and is used for cross-repository
+dependencies. For example, when `gate` code depends on `fiat` code, the jars for
+that dependency are generated by these builds. This system is also used by the
+plugin system for versioning the plugin APIs.
+
+The other system is used for the artifacts we distribute to the community:
+Docker containers and Debian packages. This system runs in Jenkins and is a
+complicated mash-up of Jenkins and Google Cloud Build configuration, shell
+scripts, Python, and Kotlin. Some of it is checked-in and some of it just lives
+in Jenkins configs or files on the Jenkins server.
+
+These build systems are mostly unaware of each other. They use different version
+numbering schemes, but with identically-shaped version numbers. For example, in
+the past year, they've each generated an echo 2.12.0, but those two 2.12.0s
+aren't related to one another in any way. Their git tags are slightly different
+(`v2.12.0` vs. `version-2.12.0`) and their artifacts are published to different
+Bintray repositories, so there are no conflicts. (Yes, we also have two
+different Bintray repositories.)
+
+## Timeline
+
+This build system will be implemented between the 1.22 release in mid-August and
+the following release (which will no longer be called 1.23) in mid-October.
+
+## Design
+
+### New Versioning Scheme and Release Schedule
+
+Spinnaker releases will switch to date-based versions, so instead of 1.23, we
+will call it 2020.4. The release schedule will change from every 8 weeks to
+every 3 months.
+
+Anecdotally, we believe that some users expect that there won't be any breaking
+changes between 1.20 and 1.21, perhaps confusing Spinnaker versions with
+[SemVer](https://semver.org/). In fact, we regularly introduce breaking changes
+in these releases.
```

Comment:
Have we ever explicitly stated that release versions aren't SemVer compliant (outside of this doc and calls)? What's the preventing us from following SemVer releases? If it's regularly introducing breaking changes, might there be a better way to solve that problem instead of increasing release cadence? Better testing? Incrementing the major version when appropriate?

Reply:
It looks like Kubernetes has a pretty [well documented](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/release/versioning.md#kubernetes-release-versioning) approach to how they version multiple components using SemVer.

Reply:
I agree that the Spinnaker version number has never actually signaled semver changes, but I don't think that's a good reason to switch to a different versioning scheme. Users want to know about breaking changes. Can we leave a TODO here and decide, at a later date and in a separate RFC, the scope of Spinnaker's public API and the definition of a breaking change w.r.t semver?

Reply:
> Have we ever explicitly stated that release versions aren't SemVer compliant

What if you flip that around: have we ever explicitly stated that release versions _are_ SemVer compliant? I doubt it. I obviously wasn't there at the time, but I don't think these version numbers were ever intended to be used that way. I think it was more like a "Microsoft Word 5.2"-style version number.

Like Dan points out, I think we'd have to have a documented API before we could even _try_ to implement SemVer. Right now I don't even know what it means. If we remove a configuration option, for example, is that considered a breaking change? If we remove Oracle support that's used by almost no one, do we have to bump the major version?

Honestly, it just doesn't seem like Spinnaker was carefully designed enough to make promises of backwards compatibility. It's a lot of very difficult work to make a promise like that. There are dozens of obscure features implemented with little to no documentation or tests, along with zero metrics to know who is using them, and to hope that we can move the product forward without breaking or removing some of them just doesn't feel realistic to me.

Or maybe we just say "everything right now is beta and we'll mark specific hand-selected things as within the scope of SemVer going forward", but how much value would that actually provide?

> I agree that the Spinnaker version number has never actually signaled semver changes, but I don't think that's a good reason to switch to a different versioning scheme.

I also listed a few other reasons :grin: But yeah, as I mentioned, we don't _have_ to do this now... I'm still inclined to push for it, but not especially vigorously.

Reply:
> it just doesn't seem like Spinnaker was carefully designed enough to make promises of backwards compatibility. 

I don't think this can be entirely true 😬, though I understand the sentiment. You could start with a small public API scope and expand over time as test coverage improves and users tell us what they care about. I'd love if we could say that pipeline configs are executable across minor releases (with a strong understanding that it's not going to be perfect, a culture of adding tests when you inevitably break minor versions, and a system for marking stage types and cloud providers as alpha/beta/not-covered-by-semver).

Reply:
> What if you flip that around: have we ever explicitly stated that release versions are SemVer compliant? 

Fair point. I think the answer is no. I _was_ surprised to hear it said out loud the first time as were a few other at Armory and from the community, though. I guess the assumption I made was that "we follow SemVer, just poorly". 


> It's a lot of very difficult work to make a promise like that.

Totally. I can't help but wonder if it's valuable work, though. If anything, understanding all of the points standing in our way - along with the points you mentioned - would give us a clear understanding of how much of a lift it is and whether it's even possible. I do agree with @danielpeach that we should define _some_ degree of backwards compatibility, however big or small it ends up being. Obviously that's outside the scope of this PR, and probably an exercise for the TOC. I'm just exploring the problem a bit. 

I'm happy to provide input/feedback on what that might look like. We (Armory) have a good bit of anecdotal data about how breaking changes affect users and the work they/we have to do to workaround them.

Reply:
I think it’s worth sharing any data (anecdotal or not). We should create space for an initial discussion on backwards compatibility expectations in the TOC.